bench-tps/src/log_transaction_service.rs

```diff
@@ -0,0 +1,498 @@
+//! `LogTransactionService` requests confirmed blocks, analyses transactions submitted by bench-tps,
+//! and saves log files in csv format.
+
+use {
+    crate::bench_tps_client::{BenchTpsClient, Result},
+    chrono::{DateTime, TimeZone, Utc},
+    crossbeam_channel::{select, tick, unbounded, Receiver, Sender},
+    log::*,
+    serde::Serialize,
+    solana_client::rpc_config::RpcBlockConfig,
+    solana_measure::measure::Measure,
+    solana_sdk::{
+        clock::{DEFAULT_MS_PER_SLOT, DEFAULT_S_PER_SLOT, MAX_PROCESSING_AGE},
+        commitment_config::{CommitmentConfig, CommitmentLevel},
+        signature::Signature,
+        slot_history::Slot,
+    },
+    solana_transaction_status::{
+        option_serializer::OptionSerializer, EncodedTransactionWithStatusMeta, RewardType,
+        TransactionDetails, UiConfirmedBlock, UiTransactionEncoding, UiTransactionStatusMeta,
+    },
+    std::{
+        collections::HashMap,
+        fs::File,
+        sync::Arc,
+        thread::{self, sleep, Builder, JoinHandle},
+        time::Duration,
+    },
+};
+
+// Data to establish communication between sender thread and
+// LogTransactionService.
+#[derive(Clone)]
+pub(crate) struct TransactionInfoBatch {
+    pub signatures: Vec<Signature>,
+    pub sent_at: DateTime<Utc>,
+    pub compute_unit_prices: Vec<Option<u64>>,
+}
+
+pub(crate) type SignatureBatchSender = Sender<TransactionInfoBatch>;
+
+pub(crate) struct LogTransactionService {
+    thread_handler: JoinHandle<()>,
+}
+
+pub(crate) fn create_log_transactions_service_and_sender<Client>(
+    client: &Arc<Client>,
+    block_data_file: Option<&str>,
+    transaction_data_file: Option<&str>,
+) -> (Option<LogTransactionService>, Option<SignatureBatchSender>)
```

Comment:
I return pair of options instead of option of pair because I will need to pass these values independently to different functions.